### PR TITLE
Integrate exoscale integration with Project Syn

### PR DESCRIPTION
**Pull request by ccremer**
* Adjusts installation instructions to VSHN specific OCP4 installations on Exoscale
* Enables OCP4 installations with/for Project Syn
* Updates decommissioning instructions

### CODE REVIEW

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 1]
```diff
@@ -70,23 +137,116 @@ EOF
...
+
+.Store various secrets in Vault
+[source,console]
+----
+# Set the exoscale.ch access secrets
+vault kv put clusters/kv/${TENANT_ID}/${CLUSTER_ID}/exoscale \
+  api_key=${EXOSCALE_API_KEY} \
+  api_secret=${EXOSCALE_API_SECRET} \
+  s3_access_key=${EXOSCALE_API_KEY} \
+  s3_secret_key=${EXOSCALE_API_SECRET}
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:203]: Why is this necessary?

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 2]
```diff
@@ -205,27 +408,32 @@ We provision the VMs in this two-step process to avoid VMs getting stuck in a st
 +
 [NOTE]
 ====
-We currently need two runs of `terraform apply -var X_state=Running` when provisioning VMs.
+We currently need two runs of `terraform apply` with the state "Running" when provisioning VMs.
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:411]:
```suggestion
We currently need two runs of `terraform apply` with state "Running" when provisioning VMs.
```

File: docs/modules/ROOT/partials/setup_terraform_exoscale.adoc [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+# Generate the terraform alias
+alias terraform='docker run -it --rm \
+  -u $(id -u) \
+  -e EXOSCALE_API_KEY \
+  -e EXOSCALE_API_SECRET \
+  -e TF_VAR_lb_exoscale_api_key \
+  -e TF_VAR_lb_exoscale_api_secret \
+  -e AWS_ACCESS_KEY_ID \
+  -e AWS_SECRET_ACCESS_KEY \
+  -e AWS_S3_ENDPOINT \
```
Comment (simu) [inline docs/modules/ROOT/partials/setup_terraform_exoscale.adoc:24]: These are not required when following the rest of the guide to use the GitLab backend

File: docs/modules/ROOT/partials/setup_terraform_exoscale.adoc [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  yq eval ".parameters.openshift4_terraform.images.terraform.tag" \
+  dependencies/openshift4-terraform/class/defaults.yml)
+
+# Generate the terraform alias
+alias terraform='docker run -it --rm \
+  -u $(id -u) \
+  -e EXOSCALE_API_KEY \
+  -e EXOSCALE_API_SECRET \
+  -e TF_VAR_lb_exoscale_api_key \
+  -e TF_VAR_lb_exoscale_api_secret \
```
Comment (simu) [inline docs/modules/ROOT/partials/setup_terraform_exoscale.adoc:21]: may be worth trying to have unified instructions for Cloudscale/Exoscale by using a environment file (docker defaults to `.env` IIRC).

**Review by simu — CHANGES_REQUESTED** [review]
Looks mostly ok, I've left inline comments for stuff which didn't work correctly out of the box when following the instructions.

File: docs/modules/ROOT/partials/setup_terraform_exoscale.adoc (hunk 4 above)
Comment (ccremer) [reply docs/modules/ROOT/partials/setup_terraform_exoscale.adoc:21]: docker-compose certainly uses .env by default. I'm not sure about `docker run`, the reference doesn't mention it here: https://docs.docker.com/engine/reference/commandline/run/#set-environment-variables--e---env---env-file.

but can be made explicitly with `--env-file .env` ofc.

Comment (simu) [reply docs/modules/ROOT/partials/setup_terraform_exoscale.adoc:21]: Ah yeah, that's right, `docker run` doesn't have an environment file which gets picked up by default, you need to explicitly specify `--env-file`. I've got no hard feelings about the name of the file itself.

Comment (ccremer) [reply docs/modules/ROOT/partials/setup_terraform_exoscale.adoc:21]: The cloudscale instructions have snippets that reference files and dirs from the openshift4-cloudscale component. I suggest to migrate to the generic component in a separate PR resp. with the official deprecation of openshift4-cloudscale

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 5]
```diff
@@ -97,65 +257,113 @@ pullSecret: |
...
 +
 [source,console]
 ----
-RHCOS_VERSION="4.7.7"
+mkdir -p inventory/classes/
+git clone $(curl -sH"Authorization: Bearer ${COMMODORE_API_TOKEN}" "${COMMODORE_API_URL}/tenants/${TENANT_ID}" | jq -r '.gitRepo.url') inventory/classes/${TENANT_ID}
+----
 
-curl "https://mirror.openshift.com/pub/openshift-v4/dependencies/rhcos/4.7/${RHCOS_VERSION}/rhcos-${RHCOS_VERSION}-x86_64-openstack.x86_64.qcow2.gz" | gunzip > rhcos-${RHCOS_VERSION}.qcow2
+. Prepare Terraform cluster config
```
[the remark this answers is not in the document]
Comment (ccremer) [reply docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:308]: Maybe we can set the bootstrap bucket via Commodore defaults for exoscale?
Like
```yaml
parameters:
  openshift4_terraform:
    terraform_variables:
      bootstrap_bucket: https://sos-${facts:region}.exo.io/${cluster:name}-bootstrap
```

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc (hunk 1 above)
Comment (ccremer) [reply docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:203]: The S3 credentials are also in the cloudscale instructions. Could it be for K8up cluster-backup?

Comment (simu) [reply docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:203]: The S3 ones are probably for K8up, but we definitely don't need `api_key` and `api_secret` as those are used by cloudscale-csi.

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 6]
```diff
@@ -97,65 +257,113 @@ pullSecret: |
...
+yq eval -i ".parameters.openshift4_terraform.terraform_variables.base_domain = \"${BASE_DOMAIN}\"" \
+  ${CLUSTER_ID}.yml
+
+yq eval -i ".parameters.openshift4_terraform.terraform_variables.ignition_ca = \"${CA_CERT}\"" \
+  ${CLUSTER_ID}.yml
+
+yq eval -i ".parameters.openshift4_terraform.terraform_variables.rhcos_template = \"${TF_VAR_rhcos_template}\"" \
+  ${CLUSTER_ID}.yml
+
+yq eval -i ".parameters.openshift4_terraform.terraform_variables.ssh_key = \"$(cat ${SSH_PUBLIC_KEY})\"" \
```
[the remark this answers is not in the document]
Comment (ccremer) [reply docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:339]: You sure about that? the content of SSH_PUBLIC_KEY in my case had the comment, which propagated to Terraform in `main.tf.json`, and I could connect to the bootstrap and master nodes just fine...?

[the remark this answers is not in the document]
Comment (simu) [reply docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:339]: Maybe Exoscale is not happy about comments with spaces, my current key has comment "simon.gerber@vshn.ch 2021".

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 7]
```diff
@@ -14,34 +14,99 @@ This how-to guide is still a work in progress and will change.
...
+.OpenShift configuration
+[source,console]
+----
+export BASE_DOMAIN=<your-base-domain>
+export PULL_SECRET='<redhat-pull-secret>' # As copied from https://cloud.redhat.com/openshift/install/pull-secret "Copy pull secret". value must be inside quotes.
+----
+
+For `BASE_DOMAIN` explanation, see xref:explanations/dns_scheme.adoc[DNS Scheme].
+
+.Exoscale configuration
```
[the remark this answers is not in the document]
Comment (ccremer) [reply docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:107]: I tried something with a recent commit. However, I'm not sure how we set the terraform variable ssh_key with vault. If we store the pubkey as a vault reference instead of plain text, would catalog compilation resp. `terraform apply` work in the output dir? Or do we expose a variable like the `lb_exoscale_api_secret` and set it with `TF_VAR_...` ?

[the remark this answers is not in the document]
Comment (simu) [reply docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:107]: > If we store the pubkey as a vault reference instead of plain text, would catalog compilation resp. `terraform apply` work in the output dir? Or do we expose a variable like the `lb_exoscale_api_secret` and set it with `TF_VAR_...` ?

We don't need to have a Vault reference for the public key, we just need to save the private key in Vault, so it's available for debugging purposes.

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 8]
```diff
@@ -275,11 +494,44 @@ oc get csr -o go-template='{{range .items}}{{if not .status}}{{.metadata.name}}{
...
+
+kubectl get nodes
+
+kubectl label --overwrite node -lnode-role.kubernetes.io/worker \
+  node-role.kubernetes.io/app=""
+kubectl label node -lnode-role.kubernetes.io/infra \
+  node-role.kubernetes.io/app-
+----
+
+. Create secret with S3 credentials https://docs.openshift.com/container-platform/4.7/registry/configuring_registry_storage/configuring-registry-storage-aws-user-infrastructure.html#registry-operator-config-resources-secret-aws_configuring-registry-storage-aws-user-infrastructure[for the registry]
```
[the remark this answers is not in the document]
Comment (ccremer) [reply docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:551]: Is it ok to use a dedicated key for S3, but for both registry and k8up? Or should they need separate keys as well?

[the remark this answers is not in the document]
Comment (simu) [reply docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:551]: Probably fine to have a shared key for registry and K8up for now. Might be worth to file a follow-up to fully figure out which IAM keys are required for OCP4 on Exoscale.

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 9]
```diff
@@ -70,23 +134,131 @@ EOF
...
 [source,console]
 ----
-exo sos create "${CLUSTER_ID}-bootstrap"
-exo sos create "${CLUSTER_ID}-tf-state"
+SSH_PRIVATE_KEY="$(pwd)/ssh_$CLUSTER_ID"
+export SSH_PUBLIC_KEY="${SSH_PRIVATE_KEY}.pub"
+
+ssh-keygen -C "vault@$CLUSTER_ID" -t ed25519 -f $SSH_PRIVATE_KEY -N ''
+
+export SSH_PUBKEY_FINGERPRINT=$(ssh-keygen -E md5 -lf $SSH_PUBLIC_KEY | cut -d' ' -f2 | tail -c+5)
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:246]: Not needed anymore, since we're now ensuring that we are uploading a public key which doesn't exist yet in the Exoscale account

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 10]
```diff
@@ -97,105 +269,149 @@ pullSecret: |
...
+  ${CLUSTER_ID}.yml
+
+yq eval -i ".parameters.openshift4_terraform.terraform_variables.rhcos_template = \"${TF_VAR_rhcos_template}\"" \
+  ${CLUSTER_ID}.yml
+
+yq eval -i ".parameters.openshift4_terraform.terraform_variables.bootstrap_bucket = \"${BOOTSTRAP_BUCKET}\"" \
+  ${CLUSTER_ID}.yml
+
+yq eval -i ".parameters.openshift4_terraform.terraform_variables.ssh_key = \"$(cut -d' ' -f1,2 <${SSH_PUBLIC_KEY})\"" \
+  ${CLUSTER_ID}.yml
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:350]: Might be fine to just have `$(cat $SSH_PUBLIC_KEY)` now that we control the comment

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 11]
```diff
@@ -70,23 +134,131 @@ EOF
...
+  --checksum $(md5sum rhcos-${RHCOS_VERSION}.qcow2 | awk '{ print $1 }') \
+  --boot-mode uefi \
+  --disable-password \
+  --username core \
+  --description "Red Hat Enterprise Linux CoreOS (RHCOS) ${RHCOS_VERSION}" \
+  --url "https://${AWS_S3_ENDPOINT}/${CLUSTER_ID}-bootstrap/rhcos-${RHCOS_VERSION}.qcow2"
+
+exo storage delete "sos://${CLUSTER_ID}-bootstrap/rhcos-${RHCOS_VERSION}.qcow2"
+
+export TF_VAR_rhcos_template="rhcos-${RHCOS_VERSION}"
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:172]:
```suggestion
export RHCOS_TEMPLATE="rhcos-${RHCOS_VERSION}"
```

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 12]
```diff
@@ -97,105 +269,149 @@ pullSecret: |
...
 
-export TF_VAR_rhcos_template="rhcos-${RHCOS_VERSION}"
+yq eval -i ".parameters.openshift4_terraform.terraform_variables.base_domain = \"${BASE_DOMAIN}\"" \
+  ${CLUSTER_ID}.yml
+
+yq eval -i ".parameters.openshift4_terraform.terraform_variables.ignition_ca = \"${CA_CERT}\"" \
+  ${CLUSTER_ID}.yml
+
+yq eval -i ".parameters.openshift4_terraform.terraform_variables.rhcos_template = \"${TF_VAR_rhcos_template}\"" \
+  ${CLUSTER_ID}.yml
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:344]:
```suggestion
yq eval -i ".parameters.openshift4_terraform.terraform_variables.rhcos_template = \"${RHCOS_TEMPLATE}\"" \
  ${CLUSTER_ID}.yml
```

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 13]
```diff
@@ -97,105 +269,149 @@ pullSecret: |
...
+include::partial$setup_terraform_exoscale.adoc[]
+
+
 . Initialize Terraform
 +
 [NOTE]
 ====
 Exoscale doesn't allow uploading the same public key multiple times in a single organisation.
 If the public key in `$SSH_PUBLIC_KEY` already exists in the organisation, we instruct Terraform to use the existing keypair instead of trying to upload the public key under a new name.
 ====
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:395]: This note can be deleted now that we generate a unique SSH key for the cluster

**Review by simu — CHANGES_REQUESTED** [review]
I noticed a few more little things that need to be adjusted.

Note: I've not re-tested the instructions using the latest state of the PR yet.

**Comment by ccremer** [comment]
I shall try another run later

**Comment by ccremer** [comment]
I retested everything just now. worked fine with latest smaller fixes.

I have also updated the decommission how-to. In there, I'm assuming a fresh, empty directory and now require a catalog compilation to configure terraform correctly.

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 14]
```diff
@@ -14,46 +14,110 @@ This how-to guide is still a work in progress and will change.
...
-export AWS_ACCESS_KEY_ID=${EXOSCALE_API_KEY}
-export AWS_SECRET_ACCESS_KEY=${EXOSCALE_API_SECRET}
-export AWS_S3_ENDPOINT="sos-${EXOSCALE_REGION}.exo.io"
+# From https://git.vshn.net/profile/personal_access_tokens
+export GITLAB_TOKEN=<gitlab-api-token>
+
+# For example: https://api.syn.vshn.net
+# IMPORTANT: do NOT add a trailing `/`. Commands below will fail.
+export COMMODORE_API_URL=<lieutenant-api-endpoint>
+# See https://wiki.vshn.net/x/ngMBCg#ClusterRegistryinLieutenantSynthesizeaCluster(synfection)-Preparation
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:87]: Not sure what was supposed to be explained in more detail in the synthesize guide in the wiki. I'd say we just remove this line.

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 15]
```diff
@@ -14,46 +14,110 @@ This how-to guide is still a work in progress and will change.
...
+export COMMODORE_API_TOKEN=<lieutenant-api-token>
+----
 
-export SSH_PUBLIC_KEY=~/.ssh/id_ed25519.pub
-export SSH_PUBKEY_FINGERPRINT=$(ssh-keygen -E md5 -lf $SSH_PUBLIC_KEY | cut -d' ' -f2 | tail -c+5)
+.VSHN sepcific input
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:91]:
```suggestion
.VSHN-specific input
```

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 16]
```diff
@@ -14,46 +14,110 @@ This how-to guide is still a work in progress and will change.
...
+export AWS_S3_ENDPOINT="sos-${EXOSCALE_REGION}.exo.io"
 
 # TODO: generate separate restricted IAM key for Ursula
-export TF_VAR_lb_exoscale_api_key=${EXOSCALE_API_KEY}
-export TF_VAR_lb_exoscale_api_secret=${EXOSCALE_API_SECRET}
+export TF_VAR_lb_exoscale_api_key=<exoscale-key-for-lb>
+export TF_VAR_lb_exoscale_api_secret=<exoscale-secret-for-lb>
 ----
 
+=== Set up S3 bucket
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:119]:
```suggestion
=== Set up S3 bucket for cluster bootstrap
```

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 17]
```diff
@@ -70,23 +134,129 @@ EOF
...
+For the following steps, change into a clean directory (for example a directory in your home).
+
+[CAUTION]
+These are the only steps which aren't idempotent and have to be completed uninterrupted in one go.
+If you have to recreate the install config or any of the generated manifests you need to rerun all of the subsequent steps.
+
+[NOTE]
+--
+You can add more options to the `install-config.yaml` file.
+Have a look at the https://docs.openshift.com/container-platform/{ocp-minor-version}/installing/installing_bare_metal/installing-bare-metal.html#installation-bare-metal-config-yaml_installing-bare-metal[config example] for more information.
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:232]: The macro definition for `{ocp-minor-version}` is missing, breaking this link.

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 18]
```diff
@@ -70,23 +134,129 @@ EOF
...
+
+For example, you could change the SDN from a default value to something a customer requests due to some network requirements.
+--
+
+. Prepare SSH key
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:237]:
```suggestion
. Prepare SSH key
+
NOTE: We generate a unique SSH key pair for the cluster, because Exoscale doesn't support configuring multiple public keys for a VM.
```

File: docs/modules/ROOT/pages/how-tos/exoscale/install.adoc [hunk 19]
```diff
@@ -275,11 +496,47 @@ oc get csr -o go-template='{{range .items}}{{if not .status}}{{.metadata.name}}{
 
 kubectl get nodes
 
-openshift-install --dir ${CLUSTER_ID} \
+kubectl get nodes -l node-role.kubernetes.io/worker
+kubectl label node -l node-role.kubernetes.io/worker \
+  node-role.kubernetes.io/infra=""
+
+openshift-install --dir ${INSTALLER_DIR} \
   wait-for install-complete
```
Comment (simu) [inline docs/modules/ROOT/pages/how-tos/exoscale/install.adoc:504]: Might make sense to make this its own step

```suggestion
----

. Wait for installation to complete
+
[source,console]
----
openshift-install --dir ${INSTALLER_DIR} \
  wait-for install-complete
```

**Review by simu — APPROVED** [review]
LGTM overall, some last minor suggestions and fixes inline.
